File: market_analyzer/BacktestRange/Backtest.cs
using Application.Services.Providers.Range;
using MiniExcelLibs.Attributes;

namespace BacktestRange;

enum SideType
{
    Buy = 1,
    Sell
}

record class Position()
{
    [ExcelFormat("dd/MM/yyyy HH:mm:ss.fff")]
    public required DateTime Date { get; init; }
    public required SideType Type { get; init; }
    public required double OpenPrice { get; init; }
    public double? ClosePrice { get; private set; }
    public double CurrentPrice { get; private set; }
    public double Profit { get; private set; }
    public bool IsOpen => ClosePrice is null;

    public void Update(double price, bool closePosition)
    {
        CurrentPrice = price;

        if (closePosition)
        {
            ClosePrice = price;
        }

        if (Type is SideType.Buy)
        {
            Profit = price - OpenPrice;
        }

        if (Type is SideType.Sell)
        {
            Profit = OpenPrice - price;
        }
    }
}

class Backtest(double slippage)
{
    public double Slippage { get; } = slippage;
    public List<Brick> Bricks { get; set; } = [];
    public List<Position> Positions { get; set; } = [];
    public Position? CurrentPosition
    {
        get
        {
            var last = Positions.LastOrDefault();
            if (last is null || !last.IsOpen)
            {
                return null;
            }
            return last;
        }
    }

    public void AddBrick(Brick brick)
    {
        Bricks.Add(brick);
        Statistics();
    }

    public void AddPosition(SideType type)
    {
        var lastPosition = Positions.LastOrDefault();

        if (lastPosition?.IsOpen ?? false)
        {
            throw new InvalidOperationException("Position open.");
        }

        var brick = Bricks.Last();

        Positions.Add(new()
        {
            Date = brick.Date,
            OpenPrice = type == SideType.Buy ? brick.Close + Slippage : brick.Close - Slippage,
            Type = type
        });

        Statistics();
    }

    public void ClosePosition()
    {
        var lastPosition = Positions.Last();

        if (!lastPosition.IsOpen)
        {
            throw new InvalidOperationException();
        }

        Statistics(closePosition: true);
    }

    private void Statistics(bool closePosition = false)
    {
        var lastPosition = Positions.LastOrDefault();

        if (!(lastPosition?.IsOpen ?? false))
        {
            return;
        }

        var brick = Bricks.Last();

        var closeType = lastPosition.Type == SideType.Buy ? SideType.Sell : SideType.Buy;

        lastPosition.Update(closeType == SideType.Buy ? brick.Close + Slippage : brick.Close - Slippage, closePosition);
    }
}

File: grpc_server/terminal/Extensions/Backtest.py
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional


class SideType(Enum):
    BUY = 1
    SELL = 2


@dataclass
class Position:
    date: datetime
    type: SideType
    open_price: float
    close_price: Optional[float] = None
    current_price: float = 0.0
    profit: float = 0.0

    @property
    def is_open(self) -> bool:
        return self.close_price is None

    def update(self, price: float, close_position: bool):
        self.current_price = price

        if close_position:
            self.close_price = price

        if self.type == SideType.BUY:
            self.profit = price - self.open_price

        if self.type == SideType.SELL:
            self.profit = self.open_price - price


@dataclass
class Brick:
    date: datetime
    close: float


@dataclass
class Backtest:
    slippage: float
    bricks: List[Brick] = field(default_factory=list)
    positions: List[Position] = field(default_factory=list)

    @property
    def current_position(self) -> Optional[Position]:
        if not self.positions:
            return None
        last = self.positions[-1]
        if not last.is_open:
            return None
        return last

    def add_brick(self, brick: Brick):
        self.bricks.append(brick)
        self.statistics()

    def add_position(self, type: SideType):
        if self.current_position:
            raise ValueError("Position open.")

        brick = self.bricks[-1]
        open_price = (
            brick.close + self.slippage
            if type == SideType.BUY
            else brick.close - self.slippage
        )

        self.positions.append(
            Position(date=brick.date, type=type, open_price=open_price)
        )
        self.statistics()

    def close_position(self):
        if not self.current_position:
            raise ValueError("No open position to close.")

        self.statistics(close_position=True)

    def statistics(self, close_position: bool = False):
        last_position = self.current_position

        if not last_position:
            return

        brick = self.bricks[-1]
        close_type = (
            SideType.SELL if last_position.type == SideType.BUY else SideType.BUY
        )
        close_price = (
            brick.close + self.slippage
            if close_type == SideType.BUY
            else brick.close - self.slippage
        )

        last_position.update(close_price, close_position)
